indy incompatibility with SecurityManager for some internal instrumentation

### Description

The Security Manager smoke test fails when invokedynamic instrumentation is enabled for some of the internal instrumentation modules.

```
OpenTelemetry Javaagent failed to start
java.lang.BootstrapMethodError: call site initialization exception
      at java.lang.invoke.CallSite.makeSite(CallSite.java:341)
      at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
      at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)
      at java.lang.invoke.InnerClassLambdaMetafactory.spinInnerClass(InnerClassLambdaMetafactory.java:251)
      at java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(InnerClassLambdaMetafactory.java:194)
      at java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:304)
      at java.lang.invoke.CallSite.makeSite(CallSite.java:302)
      at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
      at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)
      at io.opentelemetry.javaagent.tooling.AgentInstaller.addSqlCommenterCustomizers(AgentInstaller.java:326)
      at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:209)
      at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:101)
      at io.opentelemetry.javaagent.tooling.AgentStarterImpl.start(AgentStarterImpl.java:103)
      at io.opentelemetry.javaagent.bootstrap.AgentInitializer$2.run(AgentInitializer.java:68)
      at io.opentelemetry.javaagent.bootstrap.AgentInitializer$2.run(AgentInitializer.java:61)
      at java.security.AccessController.doPrivileged(Native Method)
      at io.opentelemetry.javaagent.bootstrap.AgentInitializer.doPrivilegedExceptionAction(AgentInitializer.java:108)
      at io.opentelemetry.javaagent.bootstrap.AgentInitializer.execute(AgentInitializer.java:82)
      at io.opentelemetry.javaagent.bootstrap.AgentInitializer.initialize(AgentInitializer.java:60)
      at io.opentelemetry.javaagent.OpenTelemetryAgent.startAgent(OpenTelemetryAgent.java:59)
      at io.opentelemetry.javaagent.OpenTelemetryAgent.premain(OpenTelemetryAgent.java:46)
      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke(Method.java:498)
      at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
      at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class io.opentelemetry.javaagent.bootstrap.ExceptionLogger
      at io.opentelemetry.javaagent.bootstrap.IndyBootstrapDispatcher.bootstrap(IndyBootstrapDispatcher.java:48)
      at java.lang.invoke.CallSite.makeSite(CallSite.java:306)
      ... 26 more

```

This probably indicates that some of those instrumentation modules require to use `AdviceUtil#applyInlineAdvice` (or similar) to force inline instrumentation.

Forcing `isIndyModule() == false` in `internal-lambda` and `internal-reflection` modules allow to make the error go away, which means we should probably force inline instrumentation in those modules, or identify what make the general indy implementation fail in this particular case.

### How to reproduce

in `io.opentelemetry.smoketest.SecurityManagerSmokeTest`, add `OTEL_JAVAAGENT_EXPERIMENTAL_INDY` = `true` environment variable.

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java
@@ -42,6 +42,7 @@ import io.opentelemetry.javaagent.tooling.config.EarlyInitAgentConfig;
 import io.opentelemetry.javaagent.tooling.field.FieldBackedImplementationConfiguration;
 import io.opentelemetry.javaagent.tooling.field.VirtualFieldImplementationInstaller;
 import io.opentelemetry.javaagent.tooling.field.VirtualFieldImplementationInstallerFactory;
+import io.opentelemetry.javaagent.tooling.ignore.IgnoreAllow;
 import io.opentelemetry.javaagent.tooling.ignore.IgnoredClassLoadersMatcher;
 import io.opentelemetry.javaagent.tooling.ignore.IgnoredTypesBuilderImpl;
 import io.opentelemetry.javaagent.tooling.ignore.IgnoredTypesMatcher;
@@ -292,14 +293,22 @@ public class AgentInstaller {
 
     Trie<Boolean> ignoredTasksTrie = builder.buildIgnoredTasksTrie();
     InstrumentedTaskClasses.setIgnoredTaskClassesPredicate(ignoredTasksTrie::contains);
+    Trie<IgnoreAllow> ignoredClassLoadersTrie = builder.buildIgnoredClassLoadersTrie();
+    DefineClassHandler.setIgnoredClassLoadersPredicate(
+        classLoader -> {
+          if (classLoader == null) {
+            return false;
+          }
+          IgnoreAllow ignored = ignoredClassLoadersTrie.getOrNull(classLoader.getClass().getName());
+          return ignored == IgnoreAllow.IGNORE;
+        });
 
     return agentBuilder
-        .ignore(any(), new IgnoredClassLoadersMatcher(builder.buildIgnoredClassLoadersTrie()))
+        .ignore(any(), new IgnoredClassLoadersMatcher(ignoredClassLoadersTrie))
         .or(new IgnoredTypesMatcher(builder.buildIgnoredTypesTrie()))
         .or(
-            (typeDescription, classLoader, module, classBeingRedefined, protectionDomain) -> {
-              return HelperInjector.isInjectedClass(classLoader, typeDescription.getName());
-            });
+            (typeDescription, classLoader, module, classBeingRedefined, protectionDomain) ->
+                HelperInjector.isInjectedClass(classLoader, typeDescription.getName()));
   }
 
   private static void addHttpServerResponseCustomizers(ClassLoader extensionClassLoader) {

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/DefineClassHandler.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/DefineClassHandler.java
@@ -11,6 +11,7 @@ import static java.util.Collections.emptySet;
 import io.opentelemetry.javaagent.bootstrap.DefineClassHelper.Handler;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.function.Predicate;
 import org.objectweb.asm.ClassReader;
 
 public class DefineClassHandler implements Handler {
@@ -18,7 +19,13 @@ public class DefineClassHandler implements Handler {
   private static final ThreadLocal<DefineClassContextImpl> defineClassContext =
       ThreadLocal.withInitial(() -> DefineClassContextImpl.NOP);
 
+  private static Predicate<ClassLoader> ignoredClassLoaders = (classLoader) -> false;
+
   private DefineClassHandler() {}
+
+  public static void setIgnoredClassLoadersPredicate(Predicate<ClassLoader> predicate) {
+    ignoredClassLoaders = predicate;
+  }
 
   @Override
   public DefineClassContext beforeDefineClass(
@@ -27,6 +34,11 @@ public class DefineClassHandler implements Handler {
     if (classBytes == null
         || (classBytes.length == 40
             && new String(classBytes, ISO_8859_1).startsWith("J9ROMCLASSCOOKIE"))) {
+      return null;
+    }
+
+    // skip ignore class loaders, class is not going to be instrumented anyway
+    if (ignoredClassLoaders.test(classLoader)) {
       return null;
     }
 

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/DefineClassHandler.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/DefineClassHandler.java
@@ -23,7 +23,7 @@ public class DefineClassHandler implements Handler {
 
   private DefineClassHandler() {}
 
-  public static void setIgnoredClassLoadersPredicate(Predicate<ClassLoader> predicate) {
+  static void setIgnoredClassLoadersPredicate(Predicate<ClassLoader> predicate) {
     ignoredClassLoaders = predicate;
   }
 

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyBootstrap.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyBootstrap.java
@@ -117,8 +117,12 @@ public class IndyBootstrap {
     // callsite resolution needs privileged access to call Class#getClassLoader() and
     // MethodHandles$Lookup#findStatic
     return java.security.AccessController.doPrivileged(
-        (PrivilegedAction<CallSite>)
-            () -> internalBootstrap(lookup, adviceMethodName, adviceMethodType, args));
+        new PrivilegedAction<CallSite>() {
+          @Override
+          public CallSite run() {
+            return internalBootstrap(lookup, adviceMethodName, adviceMethodType, args);
+          }
+        });
   }
 
   private static CallSite internalBootstrap(

--- a/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyBootstrap.java
+++ b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/IndyBootstrap.java
@@ -117,6 +117,8 @@ public class IndyBootstrap {
     // callsite resolution needs privileged access to call Class#getClassLoader() and
     // MethodHandles$Lookup#findStatic
     return java.security.AccessController.doPrivileged(
+        // using an anonymous class here instead of lambda because using a lambda here could trigger
+        // a nested bootstrap call from lambda instrumentation which could lead to stack overflow
         new PrivilegedAction<CallSite>() {
           @Override
           public CallSite run() {

--- a/smoke-tests/src/test/java/io/opentelemetry/smoketest/SecurityManagerIndySmokeTest.java
+++ b/smoke-tests/src/test/java/io/opentelemetry/smoketest/SecurityManagerIndySmokeTest.java
@@ -1,0 +1,15 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.smoketest;
+
+class SecurityManagerIndySmokeTest extends SecurityManagerSmokeTest {
+
+  @Override
+  protected void configure(SmokeTestOptions<Integer> options) {
+    super.configure(options);
+    options.env("OTEL_JAVAAGENT_EXPERIMENTAL_INDY", "true");
+  }
+}

--- a/smoke-tests/src/test/java/io/opentelemetry/smoketest/SecurityManagerSmokeTest.java
+++ b/smoke-tests/src/test/java/io/opentelemetry/smoketest/SecurityManagerSmokeTest.java
@@ -5,6 +5,8 @@
 
 package io.opentelemetry.smoketest;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 import org.junit.jupiter.api.condition.DisabledIf;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
@@ -26,8 +28,10 @@ class SecurityManagerSmokeTest extends AbstractSmokeTest<Integer> {
   @ParameterizedTest
   @ValueSource(ints = {8, 11, 17, 21}) // Security Manager removed in Java 25
   void securityManagerSmokeTest(int jdk) {
-    start(jdk);
+    SmokeTestOutput output = start(jdk);
     testing.waitAndAssertTraces(
         trace -> trace.hasSpansSatisfyingExactly(span -> span.hasName("test")));
+    // check whether output contains stack traces
+    assertThat(output.logLines().anyMatch(l -> l.startsWith("\tat "))).isFalse();
   }
 }

